storage/series/series.go

```diff
@@ -443,41 +450,56 @@ func (s *dbSeries) Bootstrap(blocks block.DatabaseSeriesBlocks) error {
 	if s.bs == bootstrapping {
 		return errSeriesIsBootstrapping
 	}
+	if bootstrappedBlocks == nil {
+		s.bs = bootstrapped
+		return nil
+	}
 
 	s.bs = bootstrapping
-	existingBlocks := s.blocks
-
-	multiErr := xerrors.NewMultiError()
-	if blocks == nil {
-		// If no data to bootstrap from then fallback to the empty blocks map
-		blocks = existingBlocks
-	} else {
-		// Request the in-memory buffer to drain and reset so that the start times
-		// of the blocks in the buckets are set to the latest valid times
-		s.buffer.DrainAndReset()
-
-		// If any received data falls within the buffer then we emplace it there
-		min, _ := s.buffer.MinMax()
-		for tNano, block := range blocks.AllBlocks() {
-			t := tNano.ToTime()
-			if !t.Before(min) {
-				if err := s.buffer.Bootstrap(block); err != nil {
-					multiErr = multiErr.Add(s.newBootstrapBlockError(block, err))
-				}
-				blocks.RemoveBlockAt(t)
+
+	// Request the in-memory buffer to drain and reset so that the start times
+	// of the blocks in the buckets are set to the latest valid times
+	s.buffer.DrainAndReset()
+	min, _, err := s.buffer.MinMax()
+	if err != nil {
+		return err
+	}
+
+	var (
+		numBlocksMovedToBuffer = int64(0) // Debug
+		numBlocksMerged        = int64(0) // Debug
+		multiErr               = xerrors.NewMultiError()
+	)
+	for tNano, block := range bootstrappedBlocks.AllBlocks() {
+		t := tNano.ToTime()
+		// If there is a writable, undrained series buffer bucket then store the block
+		// there and it will be merged / drained as part of the usual lifecycle.
+		if !t.Before(min) {
+			if err := s.buffer.Bootstrap(block); err != nil {
+				multiErr = multiErr.Add(s.newBootstrapBlockError(block, err))
 			}
+			numBlocksMovedToBuffer++
+			continue
 		}
 
-		// If we're overwriting the blocks then merge any existing blocks
-		// already drained
-		for _, existingBlock := range existingBlocks.AllBlocks() {
-			s.mergeBlockWithLock(blocks, existingBlock)
+		// If we're unable to put the blocks in an active series buffer bucket, then store them
+		// in the series block, merging with any existing blocks if necessary. There could be an
+		// existing block in the situation that currentTime > blockStart.Add(blockSize).Add(bufferPast),
+		// in which case the series buffer buckets may have been drained and rotated into a block, but
+		// still exist in memory because a flush hasn't occurred yet.
+		err := s.mergeBlockWithLock(block)
+		if err != nil {
+			return err
 		}
+		numBlocksMerged++
 	}
 
-	s.blocks = blocks
-	s.bs = bootstrapped
+	// Debug
+	scope := s.opts.InstrumentOptions().MetricsScope().SubScope("series-bootstrap")
+	scope.Counter("blocks-to-buffer").Inc(numBlocksMovedToBuffer)
+	scope.Counter("blocks-merged").Inc(numBlocksMerged)
```

Comment:
Do you want to keep these debug statements?

Reply:
yeah I think they're useful

Reply:
(moved to shard.go)